### PR TITLE
Missing value matrix heatmap

### PR DESCRIPTION
**Pull request by thinkh**
Closes Caleydo/lineupjs#366

**LineUp**
image

**Overview**

image

Artifical changes to show that the missing values are cell based:

image

### CODE REVIEW

File: src/renderer/NumbersCellRenderer.ts [hunk 1]
```diff
@@ -34,6 +37,11 @@ export default class NumbersCellRenderer extends ANumbersCellRenderer {
 
     return (ctx: CanvasRenderingContext2D, data: number[], offset: number, rowHeight: number) => {
       data.forEach((d: number, j: number) => {
+        if (isMissingValue(d)) {
+          renderMissingValue(ctx, col.getWidth(), context.rowHeight(j));
```
Comment (sgratzl) [inline src/renderer/NumbersCellRenderer.ts]: shouldn't it be `renderMissingValue(ctx, cellDimension, rowHeight, x, padding + offset)`

**Comment by mstreit** [comment]
Hm, the missing value encoding for heatmaps clutters quite a bit. Any suggestion how to resolve the problem?

**Comment by thinkh** [comment]
@mstreit we could make the dashes shorter (almost as dot) so that we have some whitespace between each cell.

**Comment by mstreit** [comment]
@think, yes that might help.

**Comment by thinkh** [comment]
Shorter dashes would look like this:

image

Overview:

image

In my opinion it draws a lot of attention especially in the overview mode. 

@sgratzl @mstreit What do you think?

**Comment by thinkh** [comment]
Shorter looks even better: 

image

image

**Comment by thinkh** [comment]
The last commit contains the shortest dash (a square). If you like it you can go ahead and merge the PR.

File: src/renderer/NumbersCellRenderer.ts [hunk 2]
```diff
@@ -18,9 +20,10 @@ export default class NumbersCellRenderer extends ANumbersCellRenderer {
         forEachChild(row, (d, i) => {
           const v = data[i];
           attr(<HTMLDivElement>d, {
-            title: NumbersColumn.DEFAULT_FORMATTER(v)
+            'title': NumbersColumn.DEFAULT_FORMATTER(v),
+            'class': isMissingValue(v) ? 'lu-missing' : ''
           }, {
-            'background-color': colorScale(v)
+            'background-color': isMissingValue(v) ? 'transparent' : colorScale(v)
```
Comment (sgratzl) [inline src/renderer/NumbersCellRenderer.ts]: why not `null` to remove the attribute?

File: src/renderer/NumbersCellRenderer.ts [hunk 3]
```diff
@@ -18,9 +20,10 @@ export default class NumbersCellRenderer extends ANumbersCellRenderer {
         forEachChild(row, (d, i) => {
           const v = data[i];
           attr(<HTMLDivElement>d, {
-            title: NumbersColumn.DEFAULT_FORMATTER(v)
+            'title': NumbersColumn.DEFAULT_FORMATTER(v),
```
Comment (sgratzl) [inline src/renderer/NumbersCellRenderer.ts]: escaping is not needed

File: src/renderer/NumbersCellRenderer.ts (hunk 2 above)
Comment (thinkh) [reply src/renderer/NumbersCellRenderer.ts]: True

**Comment by thinkh** [comment]
@sgratzl I addressed your changes.

**Pull request merged**